### PR TITLE
webview  onScroll and android setTextZoom

### LINKED ISSUE #197: [Android] Add setTextZoom property
Android  webview need setTextZoom，need help

### PR DESCRIPTION
**Pull request by meiqi1992**
webview  onScroll and android setTextZoom

Fixes #197.

### CODE REVIEW

**Comment by jamonholmgren** [comment]
Thank you @meiqi1992 !

1. Can you remove the non-English comments?
2. Can you update `./docs` too?

It's a fairly large PR so we'll need to get a few eyes on this. @Titozzz @empyrical

**Comment by Titozzz** [comment]
This is good work, once you add what @jamonholmgren asked for we can probably merge this

**Comment by meiqi1992** [comment]
@had update

**Comment by jamonholmgren** [comment]
I fixed merge conflicts but it looks like we still have some type errors. Can you fix @meiqi1992 ?

```

Cannot call `onScroll` with `event` bound to `event` because:
 - property `data` is missing in object type [1] but exists in object type [2] in type argument `T` [3].
 - property `navigationType` is missing in object type [2] but exists in object type [1] in type argument `T` [3].

   js/WebView.ios.js:431:26
   431|     onScroll && onScroll(event);
                                 ^^^^^

References:
   js/WebViewTypes.js:62:53
    62| export type WebViewNavigationEvent = SyntheticEvent<WebViewNavigation>;
                                                            ^^^^^^^^^^^^^^^^^ [1]
   js/WebViewTypes.js:64:50
    64| export type WebViewMessageEvent = SyntheticEvent<WebViewMessage>;
                                                         ^^^^^^^^^^^^^^ [2]
   node_modules/react-native/Libraries/Types/CoreEventTypes.js:13:28
    13| export type SyntheticEvent<T> = $ReadOnly<{|
                                   ^ [3]

```

**Comment by meiqi1992** [comment]
already updated ! @jamonholmgren

**Comment by Titozzz** [comment]
There have been 36 commits on your master branch. I'm kinda sure they do not all concern the scope of this PR. Please split this with 1PR / feature or fix to allow us to test your work.
Please note that closing this PR does not invalidate the work you did, but in order for us to merge it we need you to help us. 😄

**Comment by jaredly** [comment]
@Titozzz 🤔 I just read through the PR, and it sure looks like all of the changes *are* related to this change.

**Comment by jamonholmgren** [comment]
You're right @jaredly.

@meiqi1992 , would you be able to resolve conflicts again, please? I think we're close to merging, but just need to give it another look.

**Comment by Titozzz** [comment]
https://github.com/react-native-community/react-native-webview/pull/414 Already addresses the textZoom part. I'll also review all the parts I'm not sure are related @jamonholmgren @jaredly

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 1]
```diff
@@ -347,6 +358,51 @@ public void callInjectedJavaScript() {
       }
     }
 
+    public void linkBridge() {
```
Comment (Titozzz) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Why is this function being moved?

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 2]
```diff
@@ -438,7 +497,7 @@ public boolean onShowFileChooser(WebView webView, ValueCallback<Uri[]> filePathC
     settings.setBuiltInZoomControls(true);
     settings.setDisplayZoomControls(false);
     settings.setDomStorageEnabled(true);
-
+    settings.setTextZoom(100);
```
Comment (Titozzz) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Already addressed

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 3]
```diff
@@ -498,6 +557,12 @@ public void onDownloadStart(String url, String userAgent, String contentDisposit
     return webView;
   }
 
+  @ReactMethod
```
Comment (Titozzz) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: same

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 4]
```diff
@@ -773,7 +838,13 @@ public void receiveCommand(WebView root, int commandId, @Nullable ReadableArray
         break;
       case COMMAND_INJECT_JAVASCRIPT:
         RNCWebView reactWebView = (RNCWebView) root;
-        reactWebView.evaluateJavascriptWithFallback(args.getString(0));
+        if(!TextUtils.isEmpty(args.getString(0))){
```
Comment (Titozzz) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: I'm not sure what this does but you should not inject javascript like to setTextZoom anyway

File: docs/Reference.md [hunk 5]
```diff
@@ -58,6 +59,7 @@ This document lays out the current public properties and methods for the React N
 - [`reload`](Reference.md#reload)
 - [`stopLoading`](Reference.md#stoploading)
 - [`injectJavaScript`](Reference.md#injectjavascriptstr)
+- [`textZoom`](Reference.md#textZoom)
```
Comment (Titozzz) [inline docs/Reference.md]: remove

File: docs/Reference.md [hunk 6]
```diff
@@ -650,6 +661,14 @@ injectJavaScript('... javascript string ...');
 
 Executes the JavaScript string.
 
+### `textZoom()`
```
Comment (Titozzz) [inline docs/Reference.md]: remove

File: ios/RNCWKWebView.m [hunk 7]
```diff
@@ -602,19 +636,55 @@ - (void)evaluateJS:(NSString *)js
 - (void)      webView:(WKWebView *)webView
   didFinishNavigation:(WKNavigation *)navigation
 {
+  if (_messagingEnabled) {
```
Comment (Titozzz) [inline ios/RNCWKWebView.m]: This is out of the scope of this PR

File: js/WebView.android.js [hunk 8]
```diff
@@ -262,6 +264,16 @@ class WebView extends React.Component<WebViewSharedProps, State> {
     );
   };
 
+  //textZoom
+  textZoom = (data: number) => {
```
Comment (Titozzz) [inline js/WebView.android.js]: this is a hack, should be removed anyway

File: js/WebView.android.js [hunk 9]
```diff
@@ -310,6 +322,12 @@ class WebView extends React.Component<WebViewSharedProps, State> {
     onMessage && onMessage(event);
   };
 
+  //onScroll
+  _onScroll = (event: any) => {
+    var {onScroll} = this.props;
```
Comment (Titozzz) [inline js/WebView.android.js]: never use var, prefer let/const

File: js/WebView.ios.js [hunk 10]
```diff
@@ -386,6 +387,14 @@ class WebView extends React.Component<WebViewSharedProps, State> {
       [data],
     );
   };
+  /**
+  * set textZoom. default is 100
+  */
+  textZoom = (data: number) => {
```
Comment (Titozzz) [inline js/WebView.ios.js]: This could be a separate PR

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 11]
```diff
@@ -211,6 +216,9 @@ protected WritableMap createWebViewEvent(WebView webView, String url) {
       event.putString("title", webView.getTitle());
       event.putBoolean("canGoBack", webView.canGoBack());
       event.putBoolean("canGoForward", webView.canGoForward());
+      //get webview height
+      event.putDouble("height", webView.getContentHeight());
```
Comment (Titozzz) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: This is another feature

File: ios/RNCUIWebView.m [hunk 12]
```diff
@@ -314,18 +348,28 @@ - (void)webViewDidFinishLoad:(UIWebView *)webView
     ];
     [webView stringByEvaluatingJavaScriptFromString:source];
   }
+    //get webview height
+    CGFloat webViewHeight=[webView.scrollView contentSize].height;
```
Comment (Titozzz) [inline ios/RNCUIWebView.m]: Another feature

File: ios/RNCUIWebView.m [hunk 13]
```diff
@@ -314,18 +348,28 @@ - (void)webViewDidFinishLoad:(UIWebView *)webView
...
+
   if (_injectedJavaScript != nil) {
     NSString *jsEvaluationValue = [webView stringByEvaluatingJavaScriptFromString:_injectedJavaScript];
 
     NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+    //set webview height
+    [event setValue: [NSNumber numberWithFloat:webViewHeight] forKey:@"height"];
```
Comment (Titozzz) [inline ios/RNCUIWebView.m]: same

File: ios/RNCUIWebView.m [hunk 14]
```diff
@@ -314,18 +348,28 @@ - (void)webViewDidFinishLoad:(UIWebView *)webView
...
     event[@"jsEvaluationValue"] = jsEvaluationValue;
 
     _onLoadingFinish(event);
   }
   // we only need the final 'finishLoad' call so only fire the event when we're actually done loading.
   else if (_onLoadingFinish && !webView.loading && ![webView.request.URL.absoluteString isEqualToString:@"about:blank"]) {
-    _onLoadingFinish([self baseEvent]);
+      //_onLoadingFinish
```
Comment (Titozzz) [inline ios/RNCUIWebView.m]: same

**Comment by jaredly** [comment]
I can start a new PR with the onScroll stuff rebased onto master

**Comment by Titozzz** [comment]
@jaredly Sure, I'd gladly merge that

**Comment by jaredly** [comment]
Here it is :) https://github.com/react-native-community/react-native-webview/pull/516

**Comment by jamonholmgren** [comment]
@meiqi1992 Can you break this PR up into separate parts so we can review each individually? There's some good features in here, so I'd like to see if we can get them integrated.
